### ISSUE TITLE
Perf tracking

### ISSUE THREAD

**Issue by jonjohnsonjr** [chainguard-dev/apko#781]
Opening this as a meta-issue to track low hanging fruit for perf wins.

```[tasklist]
### Tasks
- [ ] https://github.com/chainguard-dev/apko/issues/772
- [ ] https://github.com/chainguard-dev/apko/issues/773
- [ ] https://github.com/chainguard-dev/apko/issues/775
- [ ] https://github.com/chainguard-dev/apko/issues/776
- [ ] https://github.com/chainguard-dev/apko/issues/777
- [ ] https://github.com/chainguard-dev/apko/issues/778
- [ ] https://github.com/chainguard-dev/apko/issues/779
- [ ] https://github.com/chainguard-dev/apko/issues/780
- [ ] https://github.com/chainguard-dev/go-apk/pull/78
- [ ] https://github.com/chainguard-dev/apko/pull/793
- [ ] https://github.com/chainguard-dev/apko/issues/799
- [ ] https://github.com/chainguard-dev/go-apk/pull/85
- [ ] https://github.com/chainguard-dev/go-apk/pull/87
- [ ] https://github.com/chainguard-dev/go-apk/pull/86
- [ ] https://github.com/chainguard-dev/go-apk/pull/82
- [ ] https://github.com/chainguard-dev/apko/issues/807
```

**Comment by jonjohnsonjr** [comment]
State of the world before we start optimizing things too much:
```
$ pwd
/Users/jonjohnson/src/github.com/chainguard-images/images

$ time apko publish --arch amd64 images/go/configs/1.19.apko.yaml localhost:8081/go
...
apko publish --arch amd64 images/go/configs/1.19.apko.yaml localhost:8081/go  27.45s user 4.15s system 242% cpu 13.009 total
```
![image](https://github.com/chainguard-dev/apko/assets/17863526/0117e362-9e47-4049-a9d0-cc5925a8d16e)

**Comment by jonjohnsonjr** [comment]
After https://github.com/chainguard-dev/apko/pull/782

```
apko publish --arch amd64 images/go/configs/1.19.apko.yaml localhost:8081/go  14.16s user 3.83s system 151% cpu 11.892 total
```

Mild speedup, but huge reduction in CPU usage.

![image](https://github.com/chainguard-dev/apko/assets/17863526/cdd0a903-f973-4b47-bfd9-0ea80762e33c)

You can see the relative length of `BuildLayer` shrinking vs `buildImage`.

**Comment by jonjohnsonjr** [comment]
After https://github.com/chainguard-dev/go-apk/pull/74

```

**Comment by jonjohnsonjr** [comment]
After https://github.com/chainguard-dev/go-apk/pull/75

```
apko publish --keyring-append  --repository-append  --arch amd64    13.46s user 4.56s system 227% cpu 7.917 total
```

Shaved off a little under a second.

![image](https://github.com/chainguard-dev/apko/assets/17863526/7c3449ac-19fa-4bcf-b39c-1f58964d3b5f)

**Comment by jonjohnsonjr** [comment]
Been a while since an update...

Here's a _cold_ cache:

```
apko publish --keyring-append  --repository-append  --arch amd64    12.49s user 3.26s system 123% cpu 12.779 total
```

![image](https://github.com/chainguard-dev/apko/assets/17863526/f419cb6c-725d-4041-8d47-15f8a424d27b)

Here's warm:

```
apko publish --keyring-append  --repository-append  --arch amd64    9.03s user 1.56s system 216% cpu 4.882 total
```

![image](https://github.com/chainguard-dev/apko/assets/17863526/c067b711-6e81-4bd9-a16b-d23c2dc9d772)

**Comment by jonjohnsonjr** [comment]
Got another half second from some bufio here and there.

Cold:

```
apko publish --keyring-append  --repository-append  --arch amd64    12.20s user 2.79s system 132% cpu 11.289 total
```

![image](https://github.com/chainguard-dev/apko/assets/17863526/05b3f596-fcc3-4640-900e-48b6dd75de80)


Hot:

```
apko publish --keyring-append  --repository-append  --arch amd64    9.06s user 1.56s system 251% cpu 4.223 total
```

![image](https://github.com/chainguard-dev/apko/assets/17863526/d93e51a7-00a3-4d38-9acc-3979dc1ef814)

**Comment by jonjohnsonjr** [comment]
I have a branch that gets us down to ~3s on the hot path, but it's a bit of a dead end because it mostly just makes the work we're already doing a little bit more concurrent, which doesn't actually help that much in a build-the-world scenario.

This is HEAD:

<img width="1648" alt="image" src="https://github.com/chainguard-dev/apko/assets/17863526/8b74626b-02b3-461b-9c1c-7b7e36e97e2f">


This is my branch:

<img width="1638" alt="image" src="https://github.com/chainguard-dev/apko/assets/17863526/40ab0342-d7d3-460e-b612-f3035efb4f64">

At least in these two flamegraphs, the exact same `9.26` seconds of CPU time is getting done.

Looking at where we're spending that time...

About a third of our CPU time is in pgzip compressing the final layer. Since we're doing a parallel compression, this only takes ~850ms, so that's about the speed of light for us on a hot path:

<img width="724" alt="image" src="https://github.com/chainguard-dev/apko/assets/17863526/d2d7c2d7-65bd-4d8c-b04d-a194e3cc3b53">

![image](https://github.com/chainguard-dev/apko/assets/17863526/07dba802-296c-4ba3-b8c9-8395adc132f5)

We spend 1.5 _serially_ writing things to disk and walking the filesystem to read them back from disk:

<img width="1646" alt="image" src="https://github.com/chainguard-dev/apko/assets/17863526/3f7bb5fa-561d-4e15-85f9-888c73c07447">

![image](https://github.com/chainguard-dev/apko/assets/17863526/d8bf5770-b3ca-4712-b2b3-67c1267a6a8c)

Meanwhile, we are gunzipping the data section of each APK, so that we are paying that time 2x (just concurrently):

<img width="1662" alt="image" src="https://github.com/chainguard-dev/apko/assets/17863526/7e34ae4a-4260-4c73-ba12-bbfddb1656b6">

A bit of a surprising result is that we spent a third of a second just cleaning up the temporary directory we created:

<img width="1645" alt="image" src="https://github.com/chainguard-dev/apko/assets/17863526/2a190d3a-b570-4e06-8258-2bf622319fa5">

![image](https://github.com/chainguard-dev/apko/assets/17863526/b7334e7d-6028-44b1-be91-74024ae5e1af)

Then we spend a surprising amount of time pushing images, but that's mostly because docker desktop won't stop touching my config file 🙄 

![image](https://github.com/chainguard-dev/apko/assets/17863526/d96eaed8-c255-4c5c-8838-6ca5538f2862)

```
cat ~/.docker/config.json | jq .credsStore
"desktop"
```

If we drop that, things look a little better here:

![image](https://github.com/chainguard-dev/apko/assets/17863526/34feb2ea-f688-4735-a288-7d4c7f79c520)


The SBOM generation is still pretty slow. I'm going to see if I can shift some of that left, but I managed to cut 1/3 of it in https://github.com/chainguard-dev/apko/pull/801

Anyway, looking back at where we're spending our time, very roughly:

1s writing a bunch of files to disk
0.5s reading a bunch of files from disk
0.5s cleaning up a bunch of files from disk
1s compressing things and other stuff

I have a plan to index the data section of APKs (really, just extract the tar headers) when we download them for the first time, then use that to avoid writing everything to disk. Instead, we can figure out what we _would_ write to disk, what files would get overwritten by what _would_ be written to disk by subsequent APKs, and also what files _would_ be affected by the apko config stuff (chmod and whatnot)... then we just (in parallel) compress the subset of files from each APK that _would_ have ended up in the final layer and at the very end append all these gzipped tarballs together with a bunch of metadata we compute along the way.

It should look something like this:

~1s writing a bunch of files to disk~
~0.5s reading a bunch of files from disk~
~0.5s cleaning up a bunch of files from disk~
1s compressing things and other stuff

That last bit of compressing stuff will now happen _even more concurrently_ than with pgzip, so I'm guessing that will bring us well under a second (on a hot path).

The next step after that would be to write some fun software that takes advantage of some details in DEFLATE to much more efficiently modify/recompress the existing APK's data section, which would shrink that latency by ~4-5x and get us closer to 250ms, at which point it will make some sense to revisit where we are spending our time.

**Comment by cosmicexplorer** [comment]
So I'm not sure that we have the same performance constraints here, but you may find pantsbuild/pex#2175 interesting, especially the `medusa-zip` tool to rearrange zip files really fast at https://github.com/cosmicexplorer/medusa-zip. It's not quite the same thing as taking advantage of DEFLATE, but one extreme crime I have performed is the hackery to read out the contents of a zip archive into another one without touching the file stream at all: https://github.com/cosmicexplorer/zip/blob/94c21b77b21db4133a210f335e0671f4ea85d6a0/src/read.rs#L331-L392. The zip format was made for messing around like this; I would love to see more crimes against DEFLATE too.

**Comment by cosmicexplorer** [comment]
This isn't exactly relevant except that I happened to be working on it at the same time as the above, but in https://github.com/pypa/pip/issues/12184#issuecomment-1676002698 I demonstrate the performance impact of creating a local index for pip which gets lazily updated as it crawls dependencies. Since I believe we discussed one result of this being the *publication* of indices for positions *referencing* some other compressed targz stream, I wanted to note instead that in a related but different application, I was able to generate local indices for resources as they were crawled, amortizing that transformation per-node. I would recommend trying that approach first here if you haven't solved the problem already by now.

**Comment by cosmicexplorer** [comment]
It would also seem very much within the scope of something like [`medusa-zip`](https://github.com/cosmicexplorer/medusa-zip) to handle the creation of such indices in a streaming manner when a targz is first downloaded.

*The following is mostly a note to self:*
One additional concern that arose from the zip merging solution investigated in pantsbuild/pex#2158 and pantsbuild/pex#2175 was that (as initially proposed) merging zip files from a shared cache would also take up more disk space than before (to create the cached zips). While handling that cache is in one sense an application-specific issue (see pantsbuild/pex#2201), if we also expand this `medusa-zip` archive library's capabilities to cover targz merging via creation of decompressed indices (and therefore hand over responsibility for the lifecycle of filesystem cache entries to that library/service), we could have it handle cache eviction/etc for the local entries it handles.

I'll create a separate issue if I have further thoughts on any of this and stop derailing this thread.

**Comment by cosmicexplorer** [comment]
Although, regarding this approach in particular:

> I have a plan to index the data section of APKs (really, just extract the tar headers) when we download them for the first time, then use that to avoid writing everything to disk. Instead, we can figure out what we would write to disk, what files would get overwritten by what would be written to disk by subsequent APKs, and also what files would be affected by the apko config stuff (chmod and whatnot)... then we just (in parallel) compress the subset of files from each APK that would have ended up in the final layer and at the very end append all these gzipped tarballs together with a bunch of metadata we compute along the way.

In order to execute build processes in isolated chroots that can be cached and executed remotely via the [bazel remexec api](https://github.com/bazelbuild/remote-apis), [pants](https://github.com/pantsbuild/pants) maintains a virtual filesystem consisting of merkle trees stored in an LMDB content-addressed store, which can be efficiently synced against a remote database (since the db only contains a mapping of `(checksum) -> (byte string)`, and entries are stored as encoded protobufs). It exposes this to build tasks with a [pretty novel API](https://www.pantsbuild.org/docs/rules-api-file-system).

Your problem here can be solved without the global deduplication that pants performs, but I wanted to mention how encoding directory contents into merkle trees is a useful general approach for performing (as you said) *"figure out what we would write to disk, what files would get overwritten by what would be written to disk by subsequent APKs, and also what files would be affected by the apko config stuff (chmod and whatnot)..."*. This act of *normalization* into a db-friendly format (in pants's case, converting directory trees into protobufs referencing other entities by checksum) may be the link that lets us meaningfully generalize this into a library, one which:
1. efficiently reads/normalizes zips/targz into a local LMDB store
    - encoded into protobufs like pants
2. efficiently computes the result of superimposing/transforming a sequence of normalized directory trees,
    - without filesystem operations; this is [what pants does with `Digest` and `Snapshot`](https://www.pantsbuild.org/docs/rules-api-file-system)
3. has methods to efficiently export a normalized directory tree into zip/targz
    - not done in pants, but see e.g. pantsbuild/pants#19049 for similar optimizations

**Comment by jonjohnsonjr** [comment]
After https://github.com/chainguard-dev/go-apk/pull/98

## Cold

~11s -> 4.9s

This came mostly from being able to fetch and decompress in parallel, which speeds up the installation phase.

![image](https://github.com/chainguard-dev/apko/assets/17863526/0b801f8d-229b-4de9-8def-1e23f137ca6a)

## Hot

~4.2s -> 2.6s

We still have that faster install phase but we get to skip the fetch phase entirely.

![image](https://github.com/chainguard-dev/apko/assets/17863526/c65548c2-f461-4a03-8836-0da17462c8ea)

**Comment by jonjohnsonjr** [comment]
After https://github.com/chainguard-dev/apko/pull/860

Using `--offline` flag (can't do this cold) on the hot path saves ~200ms mostly from avoiding TLS handshake at the beginning.

![image](https://github.com/chainguard-dev/apko/assets/17863526/da552071-c07d-42c7-86a0-d60b905554eb)

**Comment by jonjohnsonjr** [comment]
With https://github.com/chainguard-dev/apko/pull/867

Building `cgr.dev/chainguard/go` for both amd64 and arm64...

## cold

4.9s -> 3.7s

We are mostly limited here by how quickly we can fetch and decompress each APK.

![image](https://github.com/chainguard-dev/apko/assets/17863526/e35d1cec-93f2-4cb1-aac4-e1df9056ea53)

## hot

2.6s -> 1.5s

We spend most of our time now in pgzip with a bit of time burned doing TLS handshakes at the beginning and SBOM generation (giant JSON document rendering) at the end.

![image](https://github.com/chainguard-dev/apko/assets/17863526/fe229b41-718f-4845-a236-ec9dbfea98e8)

### offline

2.4s -> 1.3s

![image](https://github.com/chainguard-dev/apko/assets/17863526/9068ef88-10e7-4821-8f92-0394ddc5ecb7)

**Comment by jonjohnsonjr** [comment]
The next phase is to take this (CPU) hungry hungry pgzippopotamus and replace it with something that can go faster with less CPU. I'd even be with with a slightly slower implementation that would allow use much less CPU.

There is a particularly ambitious optimization we can perform where we could stitch together pre-existing DEFLATE streams when we _know_ that their decompressed contents are identical, which would let us reuse the CPU-intensive parts of compressing all these files.

Ensuring that the decompressed contents are identical is very difficult in the general case, but we can skip that difficulty by taking advantage of APK checksums where we already do know that the contents are identical. This requires writing a custom DEFLATE encoder, which might be out of reach for the amount of time I have here, but I want to write it here for posterity in case I come back to it.

**Comment by coreydaley-cg** [comment]
All tracked items are resolved or closed. Closing the tracking issue; new perf work gets its own issue.
